ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -3917,12 +3972,26 @@ public boolean dropPartition(String dbName, String tableName, List<String> parti
   public boolean getPartitionsByExpr(Table tbl, ExprNodeGenericFuncDesc expr, HiveConf conf,
       List<Partition> result) throws HiveException, TException {
     assert result != null;
+    boolean hasUnknownParts;
     byte[] exprBytes = SerializationUtilities.serializeExpressionToKryo(expr);
     String defaultPartitionName = HiveConf.getVar(conf, ConfVars.DEFAULTPARTITIONNAME);
-    List<org.apache.hadoop.hive.metastore.api.PartitionSpec> msParts =
-        new ArrayList<>();
-    boolean hasUnknownParts = getMSC().listPartitionsSpecByExpr(tbl.getDbName(),
-        tbl.getTableName(), exprBytes, defaultPartitionName, (short)-1, msParts);
+    List<org.apache.hadoop.hive.metastore.api.PartitionSpec> msParts = new ArrayList<>();
+
+    if (AcidUtils.isTransactionalTable(tbl)) {
+      PartitionsByExprRequest req = new PartitionsByExprRequest();
+      req.setDbName(tbl.getDbName());
+      req.setTblName((tbl.getTableName()));
+      req.setDefaultPartitionName(defaultPartitionName);
+      req.setMaxParts((short) -1);
+      req.setExpr(exprBytes);
+      ValidWriteIdList validWriteIdList = getValidWriteIdList(tbl.getDbName(), tbl.getTableName());
+      req.setValidWriteIdList(validWriteIdList != null ? validWriteIdList.toString() : null);
+      hasUnknownParts = getMSC().listPartitionsSpecByExpr(req, msParts);
+    } else {
+      hasUnknownParts = getMSC()
+          .listPartitionsSpecByExpr(tbl.getDbName(), tbl.getTableName(), exprBytes, defaultPartitionName, (short) -1,
```

Comment:
Maybe use the same listPartitionsSpecByExpr for both cases?

Reply:
Makes sense, done.